engine/src/graphics/device.cpp

```diff
@@ -0,0 +1,355 @@
+// Copyright (c) 2023-present Genesis Engine contributors (see LICENSE.txt)
+
+#include "graphics/device.hpp"
+#include "graphics/vkHelpers.hpp"
+#include "logger/log.hpp"
+
+#include <GLFW/glfw3.h>
+
+namespace gen
+{
+
+	GraphicsDevice::GraphicsDevice(gen::Window & window) : m_window{window}
+	{
+		// TODO: Allow for the vulkan api to be specified in a config file or through command line arguments.
+		// We could also dynamically check for the highest supported version of the vulkan api, but that feels outside the scope of this project.
+		createInstance(window.getTitle(), "Genesis Engine", validationLayers, {}, VK_API_VERSION_1_0);
+
+	    //createDebugMessenger();
+        //createSurface();
+		//pickPhysicalDevice();
+		//createLogicalDevice();
+		//createCommandPool();
+	}
+
+	GraphicsDevice::~GraphicsDevice()
+	{
+		m_device.destroyCommandPool(m_commandPool);
+        m_device.destroy();
+
+		//NOLINTNEXTLINE validation layers are enabled conditionally and this trips up clang-tidy
+		if (enableValidationLayers) { m_instance.destroyDebugUtilsMessengerEXT(m_debugMessenger); }
+
+		m_instance.destroySurfaceKHR(m_surface);
+		m_instance.destroy();
+	}
+
+	void GraphicsDevice::createInstance(const std::string & appName, const std::string & engineName, const std::vector<std::string> & layers,
+											 const std::vector<std::string> & extensions, const gen::u32 & apiVersion)
+	{
+#if (VULKAN_HPP_DISPATCH_LOADER_DYNAMIC == 1)
+		static vk::DynamicLoader const vkdl;
+		auto vkGetInstanceProcAddr = vkdl.getProcAddress<PFN_vkGetInstanceProcAddr>("vkGetInstanceProcAddr");
+		VULKAN_HPP_DEFAULT_DISPATCHER.init(vkGetInstanceProcAddr);
+#endif
+
+		if (enableValidationLayers) //NOLINT validation layers are enabled conditionally and this trips up clang-tidy
+        {
+			if (!vk::util::checkValidationLayerSupport(validationLayers))
+			{
+				gen::logger::error("vulkan", "validation layers requested, but not available!");
+				throw std::runtime_error("validation layers requested, but not available!");
```

Comment:
I don't think this is a fatal error, the app can run fine without validation layers, just disable them.